docs/mdx-plugins/remark-link-rewrite.js

```diff
@@ -0,0 +1,100 @@
+const path = require('path');
+const visit = require('unist-util-visit');
+const { URL } = require('url');
+
+const DEFAULT_OPTIONS = {
+  extension: 'md',
+  pageDir: 'pages',
+  trailingSlash: true,
+};
+
+/**
+ * This rewrites relative MDX links to something Next would expect.
+ *   - Replaces the `.md` extension with a trailing slash.
+ *     e.g. from: /push-notifications/overview.md, to: ./using-fcm.md -> ../using-fcm/
+ *   - Rewrites to absolute path from the page directory (backwards compatibility)
+ *     e.g. from: /push-notifications/overview.md, to ../using-fcm/ -> /push-notifications/using-fcm/
+ *
+ * @param {object} options
+ * @param {string} [options.extension="md"]
+ * @param {string} [options.pageDir="pages"]
+ * @param {boolean} [options.trailingSlash=true]
+ * @returns {function} remark plugin
+ */
+module.exports = function remarkLinkRewrite(options) {
+  const settings = Object.assign({}, DEFAULT_OPTIONS, options);
+
+  return (tree, file) => {
+    // get the relative path, from pages dir to the current file
+    const filePrefix = getFilePrefix(settings, file);
+
+    visit(tree, 'link', node => {
+      if (!isFullUrl(node)) {
+        // remove the `.md` or replace it with a trailing `/`
+        node.url = getPathWithoutExt(settings, node.url);
+        // if the path is relative, we need to clean it for Next
+        node.url = getPathRelativeUrl(filePrefix, node.url);
+      }
+    });
+  };
+};
+
+/**
+ * Remove the extension from the path, if it has one.
+ * It removes or replaces it with a trailing slash, if that's enabled.
+ */
+function getPathWithoutExt(options, url) {
+  return url.replace(`.${options.extension}`, options.trailingSlash ? '/' : '');
+}
+
+/**
+ * If the path starts with either `./` or `../` we need to clean this.
+ * Next doesn't work nicely with these kinds of paths.
+ *   - from: /push-notifications/overview/, to: sending-notifications/ -> sending-notifications/
+ *   - from: /push-notifications/overview/, to: ./sending-notifications/ -> /push-notifications/sending-notifications/
```

Comment:
this particular case seems odd to me. it may end up to be confusing that `./sending-notifications/` and `../sending-notifications/` are both resolved to the same path `/push-notifications/sending-notifications/`. maybe we need to update all occurrences of one of these cases.

Reply:
Yes, this is the confusing backwards compatibility with hardcoded trailing slash. I'll continue on this PR to remove that behavior, and add a stacked PR to change the current usage. The rewriter will basically be the same behavior Github has when linking to files directly, that makes it so much more understandable.